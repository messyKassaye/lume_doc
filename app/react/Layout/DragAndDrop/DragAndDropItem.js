import { DragSource, DropTarget } from 'react-dnd';
import { findDOMNode } from 'react-dom';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';

import { Icon } from 'UI';

export const itemSource = {
  beginDrag(props) {
    return {
      id: props.id,
      name: props.name,
      items: props.items,
      container: props.container,
      index: props.index,
      originalItem: props.originalItem,
    };
  },

  endDrag(props, monitor) {
    const container = monitor.getDropResult();
    if (container && container.id !== props.container.id) {
      props.removeItem(props.id);
    }
  },
};

export const itemTarget = {
  hover(props, monitor, component) {
    const item = monitor.getItem();
    const dragIndex = item.index;
    const hoverIndex = props.index;

    if (props.id === item.id) {
      return;
    }

    if (props.container.id !== item.container.id) {
      return;
    }

    // Don't replace items with themselves
    if (dragIndex === hoverIndex) {
      return;
    }

    // Determine rectangle on screen
    const hoverBoundingRect = findDOMNode(component).getBoundingClientRect(); // eslint-disable-line react/no-find-dom-node

    // Get vertical middle
    const hoverMiddleY = (hoverBoundingRect.bottom - hoverBoundingRect.top) / 2;

    // Determine mouse position
    const clientOffset = monitor.getClientOffset();

    // Get pixels to the top
    const hoverClientY = clientOffset.y - hoverBoundingRect.top;

    // Only perform the move when the mouse has crossed half of the items height
    // When dragging downwards, only move when the cursor is below 50%
    // When dragging upwards, only move when the cursor is above 50%

    // Dragging downwards
    if (dragIndex < hoverIndex && hoverClientY < hoverMiddleY) {
      return;
    }

    // Dragging upwards
    if (dragIndex > hoverIndex && hoverClientY > hoverMiddleY) {
      return;
    }

    // Time to actually perform the action
    props.moveItem(dragIndex, hoverIndex, item);

    // Note: we're mutating the monitor item here!
    // Generally it's better to avoid mutations,
    // but it's good here for the sake of performance
    // to avoid expensive index searches.
    item.index = hoverIndex;
    return null;
  },
};

export class DragAndDropItem extends Component {
  render() {
    const {
      iconHandle,
      isDragging,
      connectDragPreview,
      connectDragSource,
      connectDropTarget,
    } = this.props;
    let propertyClass = 'list-group-item';
    if (isDragging) {
      propertyClass += ' dragging';
    }

    if (!iconHandle) {
      propertyClass += ' draggable';
    }

    const result = connectDropTarget(
      <div className={propertyClass}>
        {iconHandle ? (
          connectDragSource(
            <span className="draggable">
              <Icon icon="bars" />
            </span>
          )
        ) : (
          <Icon icon="bars" />
        )}
        {this.props.children(this.props.originalItem, this.props.index)}
      </div>
    );

    if (!iconHandle) {
      return connectDragSource(result);
    }

    return connectDragPreview(result);
  }
}

DragAndDropItem.defaultProps = {
  iconHandle: false,
  children: () => {},
};

DragAndDropItem.propTypes = {
  connectDragSource: PropTypes.func.isRequired,
  connectDragPreview: PropTypes.func.isRequired,
  connectDropTarget: PropTypes.func.isRequired,
  index: PropTypes.number.isRequired,
  isDragging: PropTypes.bool.isRequired,
  iconHandle: PropTypes.bool,
  id: PropTypes.any.isRequired,
  moveItem: PropTypes.func.isRequired,
  children: PropTypes.func,
  originalItem: PropTypes.object.isRequired,
};

let dragAndDropItem = DropTarget('DRAG_AND_DROP_ITEM', itemTarget, connector => ({
  connectDropTarget: connector.dropTarget(),
}))(DragAndDropItem);

dragAndDropItem = DragSource('DRAG_AND_DROP_ITEM', itemSource, (connector, monitor) => ({
  connectDragSource: connector.dragSource(),
  connectDragPreview: connector.dragPreview(),
  isDragging: monitor.isDragging(),
}))(dragAndDropItem);

export default connect()(dragAndDropItem);
