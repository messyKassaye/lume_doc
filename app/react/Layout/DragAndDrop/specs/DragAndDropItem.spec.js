import React from 'react';
import { shallow } from 'enzyme';
import { DragAndDropItem } from '../DragAndDropItem';

describe('DragAndDropItem', () => {
  describe('react component', () => {
    let component;
    let props;
    beforeEach(() => {
      props = {
        connectDragSource: jasmine.createSpy('connectDragSource').and.callFake(value => value),
        connectDropTarget: jasmine.createSpy('connectDropTarget').and.callFake(value => value),
        connectDragPreview: jasmine.createSpy('connectDragPreview'),
        index: 1,
        isDragging: false,
        id: 1,
        originalItem: { label: 'item1' },
        moveItem: jasmine.createSpy('moveItem'),
        children: (item, index) => (
          <span>
            Showing item {item.label} {index}
          </span>
        ),
      };
    });

    const render = () => {
      component = shallow(<DragAndDropItem {...props} />);
    };

    it('should render item using children prop as render function', () => {
      render();
      expect(component).toMatchSnapshot();
    });

    it('should connect the component to the dragSource and dragTarget', () => {
      render();
      expect(props.connectDragSource).toHaveBeenCalled();
      expect(props.connectDropTarget).toHaveBeenCalled();
    });

    describe('when is dragging', () => {
      it('should add the class dragging to it', () => {
        props.isDragging = true;
        render();
        expect(component.find('div').hasClass('dragging')).toBe(true);
      });
    });
  });
});
