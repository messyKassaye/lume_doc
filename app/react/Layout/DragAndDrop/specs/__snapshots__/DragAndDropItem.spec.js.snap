// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DragAndDropItem react component should render item using children prop as render function 1`] = `
<div
  className="list-group-item draggable"
>
  <Connect(Icon)
    icon="bars"
  />
  <span>
    Showing item 
    item1
     
    1
  </span>
</div>
`;
